control-plane/csi-controller/src/controller.rs

```diff
@@ -0,0 +1,726 @@
+                        return Err(Status::invalid_argument(
+                            "Volume topology other than hostname not supported",
+                        ));
+                    }
+                    allowed_nodes.push(v.to_string());
+                }
+            }
+
+            for p in reqs.preferred.iter() {
+                for (k, v) in p.segments.iter() {
+                    // Ignore others than hostname (it's only preferred)
+                    if k == K8S_HOSTNAME {
+                        preferred_nodes.push(v.to_string());
+                    }
+                }
+            }
+        }
+
+        let u = Uuid::parse_str(&volume_uuid).map_err(|_e| {
+            Status::invalid_argument(format!("Malformed volume UUID: {}", volume_uuid))
+        })?;
+
+        // First check if the volume already exists.
+        if let Some(existing_volume) = MayastorApiClient::get_client()
+            .list_volumes()
+            .await?
+            .into_iter()
+            .find(|v| v.spec.uuid == u)
+        {
+            check_existing_volume(&existing_volume, replica_count, size, pinned_volume)?;
+            debug!(
+                "Volume {} already exists and is compatible with requested config",
+                volume_uuid
+            );
+        } else {
+            MayastorApiClient::get_client()
+                .create_volume(
+                    &volume_uuid,
+                    replica_count,
+                    size,
+                    &allowed_nodes,
+                    &preferred_nodes,
+                )
+                .await?;
+
+            debug!(
+                "Volume {} successfully created, pinned volume = {}",
+                volume_uuid, pinned_volume
+            );
+        }
+
+        let vt_mapper = VolumeTopologyMapper::init().await?;
+
+        let volume = rpc::csi::Volume {
+            capacity_bytes: size as i64,
+            volume_id: volume_uuid,
+            volume_context: args.parameters.clone(),
+            content_source: None,
+            accessible_topology: vt_mapper.volume_accessible_topology(pinned_volume),
+        };
+
+        debug!("Created volume: {:?}", volume);
+
+        Ok(Response::new(CreateVolumeResponse {
+            volume: Some(volume),
+        }))
+    }
+
+    async fn delete_volume(
+        &self,
+        request: tonic::Request<DeleteVolumeRequest>,
+    ) -> Result<tonic::Response<DeleteVolumeResponse>, tonic::Status> {
+        let args = request.into_inner();
+
+        debug!("Request to delete volume: {:?}", args);
+        MayastorApiClient::get_client()
+            .delete_volume(&args.volume_id)
+            .await
+            .map_err(|e| {
+                Status::internal(format!(
+                    "Failed to delete volume {}, error = {:?}",
+                    args.volume_id, e
+                ))
+            })?;
+
+        Ok(Response::new(DeleteVolumeResponse {}))
```

Comment:
do we need another debug! "deleted volume" ?

Reply:
Yes, can be added here too.